qiskit_alice_bob_provider/plugins/ab_asap.py

```diff
@@ -0,0 +1,114 @@
+
+from qiskit.transpiler import PassManager, PassManagerConfig
+from qiskit.transpiler.instruction_durations import InstructionDurations
+from qiskit.transpiler.passes import TimeUnitConversion
+from qiskit.transpiler.preset_passmanagers.common import generate_scheduling
+from qiskit.transpiler.preset_passmanagers.plugin import PassManagerStagePlugin
+
+from qiskit_alice_bob_provider.local.instruction_durations import (
+    ProcessorInstructionDurations,
+)
+
+
+class CustomTimeUnitConversion(TimeUnitConversion):
+    """
+    Choose a time unit to be used in the following time-aware passes,
+    and make all circuit time units consistent with that.
+
+    This pass will add a :attr:`.Instruction.duration` metadata to each op
+    whose duration is known which will be used by subsequent scheduling passes
+    for scheduling.
+
+    If ``dt`` (in seconds) is known to transpiler, the unit ``'dt'`` is chosen.
+    Otherwise, the unit to be selected depends on what units are used in delays
+    and instruction durations:
+    * ``'s'``: if they are all in SI units.
+    * ``'dt'``: if they are all in the unit ``'dt'``.
+    * raise error: if they are a mix of SI units and ``'dt'``.
+
+    This is a replacement of the base TimeUnitConversion pass to handle the
+    ProcessorInstructionDurations object created for Alice & Bob backends.
+
+    The pass overrides the _update_inst_durations method to apply the base
+    durations from Alice & Bob's ProcessorInstructionDuration instead of
+    the generic InstructionDurations class from Qiskit.
+    """
+
+    def _update_inst_durations(self, dag) -> InstructionDurations:
+        """Update instruction durations with circuit information.
+        If the dag contains gate calibrations and no instruction durations were
+        provided through the target or as a standalone input, the circuit
+        calibration durations will be used. The priority order for
+        instruction durations is: target > standalone > circuit.
+        """
+        if not isinstance(self.inst_durations, ProcessorInstructionDurations):
+            return self.inst_durations
+
+        # pylint: disable=protected-access
+        circ_durations = ProcessorInstructionDurations(
+            self.inst_durations._proc
+        )
+
+        if dag.calibrations:
+            cal_durations = []
+            for gate, gate_cals in dag.calibrations.items():
+                for (qubits, parameters), schedule in gate_cals.items():
+                    cal_durations.append(
+                        (gate, qubits, parameters, schedule.duration)
+                    )
+            circ_durations.update(cal_durations, circ_durations.dt)
+
+        if self._durations_provided:
+            dt = getattr(self.inst_durations, 'dt', None)
+            if isinstance(dt, float):
+                circ_durations.update(self.inst_durations, dt)
+
+        return circ_durations
+
+
+class AliceBobASAPSchedulingPlugin(PassManagerStagePlugin):
```

Comment:
Same comment here: should we open an issue (or a PR) on qiskit to allow using subclasses of TimeUnitConversion (so that we can get rid of this file) ?

Reply:
Done